Koha/REST/V1/Lists.pm

```diff
@@ -16,17 +16,323 @@ package Koha::REST::V1::Lists;
 use Koha::Virtualshelves;
-
-use Try::Tiny qw( catch try );
+use Koha::Virtualshelfcontent;
 
 =head1 API
 
 =head2 Methods
 
+=head3 list
+
+List all virtual shelves
+
+=cut
+
+sub list {
+    my $c = shift->openapi->valid_input or return;
+
+    my @lists;
+    my $virtualshelves = Koha::Virtualshelves->search();
+    while (my $list = $virtualshelves->next) {
+        push @lists, {
+            list_id       => $list->id,
+            name => $list->shelfname,
+            creation_date => $list->created_on,
+            owner_id    => $list->owner,
+            allow_change_from_others => $list->allow_change_from_others,
+            allow_change_from_owner => $list->allow_change_from_owner,
+            public   => $list->public,
+            default_sort_field => $list->sortfield,
+            records  => \@records,
+        };
+    }
+
+    return $c->render(json => \@lists);
+}
+
+=head3 create
+
+Create a virtual shelf
+
+=cut
+
+sub create {
+    my $c = shift->openapi->valid_input or return;
+    my $params = $c->req->json;
+
+    my $list = Koha::Virtualshelf->new({
+        shelfname => $params->{shelfname},
+        owner     => $params->{owner},
+        public    => $params->{public} // 0,
+    })->store;
+
+    # Add biblios to the newly created list
+    if (exists $params->{biblionumbers} && ref $params->{biblionumbers} eq 'ARRAY') {
+        foreach my $biblionumber (@{$params->{biblionumbers}}) {
+            Koha::Virtualshelfcontent->new({
+                shelfnumber  => $list->shelfnumber,
+                biblionumber => $biblionumber,
+                dateadded    => DateTime->now,
+                borrowernumber => $params->{owner},
+            })->store;
+        }
+    }
+
+    return $c->render(status => 201, json => { id => $list->id });
+}
+
+=head3 read
+
+List the contents of a virtual shelf
+
+=cut
+
+sub read {
+    my $c = shift->openapi->valid_input or return;
+    my $id = $c->param('id');
+
+    my $list = Koha::Virtualshelves->find($id);
+    return $c->render(status => 404, json => { error => 'Not found' }) unless $list;
+
+    my $records_rs = $list->get_contents;
+    my @records;
+    while (my $record = $records_rs->next) {
+        push @records, {
+            shelfnumber   => $record->shelfnumber,
+            biblionumber  => $record->biblionumber,
+            dateadded     => $record->dateadded,
+            borrowernumber => $record->borrowernumber,
+        };
+    }
+
+    return $c->render(json => {
+        list_id       => $list->id,
+        name => $list->shelfname,
+        creation_date => $list->created_on,
+        owner_id    => $list->owner,
+        allow_change_from_others => $list->allow_change_from_others,
+        allow_change_from_owner => $list->allow_change_from_owner,
+        public   => $list->public,
+        default_sort_field => $list->sortfield,
+        records  => \@records,
+    });
+}
+
+=head3 update
+
+Update a virtual shelf
+
+=cut
+
+sub update {
```

Comment:
There's another template for this:

```perl
sub update {
    my $c = shift->openapi->valid_input or return;

    my $city = Koha::Cities->find( $c->param('city_id') );

    return $c->render_resource_not_found("City")
        unless $city;

    return try {
        $city->set_from_api( $c->req->json );
        $city->store();
        return $c->render( status => 200, openapi => $c->objects->to_api($city), );
    }
    catch {
        $c->unhandled_exception($_);
    };
}
```

We'll want to think about how we handle adding and removing biblios to/from lists.. lets keep the basic crud at the list level only however.